ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs

```diff
@@ -1653,32 +1653,34 @@ withConnectionManager ConnectionManagerArguments {
             (trans, mbAssertion) <- atomically $ do
               connState <- readTVar connVar
 
-              let assertion = case connState of
-                    ReservedOutboundState ->
-                      Nothing
-                    _                     ->
-                      Just (CM.TrUnexpectedlyFalseAssertion
-                              (RequestOutboundConnection
-                                (Just connId)
-                                (abstractState (Known connState))
-                              )
-                           )
               -- @
               --  Connected : ReservedOutboundState
               --            → UnnegotiatedState Outbound
               -- @
-                  connState' = UnnegotiatedState provenance connId connThread
-              writeTVar connVar connState'
-              return ( mkTransition connState connState'
-                     , assertion
-                     )
-
-            whenJust mbAssertion $ \tr -> do
-              traceWith tracer tr
-              _ <- evaluate (assert False)
-              pure ()
+              case connState of
+                ReservedOutboundState -> do
+                  let connState' = UnnegotiatedState provenance connId connThread
+                  writeTVar connVar connState'
+                  return ( Just $ mkTransition connState connState'
+                         , Nothing
+                         )
+                TerminatingState {} ->
+                  return (Nothing, Nothing)
+                TerminatedState {} ->
+                  return (Nothing, Nothing)
+                _ -> 
+                  return ( Nothing
+                         , Just (CM.TrUnexpectedlyFalseAssertion
+                                   (RequestOutboundConnection
+                                     (Just connId)
+                                     (abstractState (Known connState))
+                                   )
+                                )
+                         )
 
-            traceWith trTracer (TransitionTrace peerAddr trans)
+            traverse_ (traceWith trTracer . TransitionTrace peerAddr) trans
+            traverse_ (traceWith tracer >=> evaluate . assert True)
+                      mbAssertion
```

Comment:
Nice, is there an open issue for this found bug?

Reply:
No, instead I just added `io-sim-discovered` flag to this PR.